adapters/apacdex/apacdex.go

```diff
@@ -84,74 +94,81 @@ func preprocess(request *openrtb2.BidRequest) error {
 	return nil
 }
 
-// MakeBids based on valueimpression server response
-func (a *ValueImpressionAdapter) MakeBids(bidRequest *openrtb2.BidRequest, unused *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
 	if responseData.StatusCode == http.StatusNoContent {
 		return nil, nil
 	}
 
 	if responseData.StatusCode == http.StatusBadRequest {
-		return nil, []error{&errortypes.BadInput{
-			Message: fmt.Sprintf("Bad user input: HTTP status %d", responseData.StatusCode),
-		}}
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
 	}
 
 	if responseData.StatusCode != http.StatusOK {
-		return nil, []error{&errortypes.BadServerResponse{
-			Message: fmt.Sprintf("Bad server response: HTTP status %d", responseData.StatusCode),
-		}}
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
 	}
 
-	var bidResponse openrtb2.BidResponse
+	var response openrtb2.BidResponse
 
-	if err := json.Unmarshal(responseData.Body, &bidResponse); err != nil {
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
 		return nil, []error{&errortypes.BadServerResponse{
 			Message: err.Error(),
 		}}
 	}
 
-	if len(bidResponse.SeatBid) == 0 {
+	if len(response.SeatBid) == 0 {
 		return nil, nil
 	}
 
-	rv := adapters.NewBidderResponseWithBidsCapacity(len(bidResponse.SeatBid[0].Bid))
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
 	var errors []error
 
-	for _, seatbid := range bidResponse.SeatBid {
+	for _, seatbid := range response.SeatBid {
 		for _, bid := range seatbid.Bid {
-			foundMatchingBid := false
-			bidType := openrtb_ext.BidTypeBanner
-			for _, imp := range bidRequest.Imp {
-				if imp.ID == bid.ImpID {
-					foundMatchingBid = true
-					if imp.Banner != nil {
-						bidType = openrtb_ext.BidTypeBanner
-					} else if imp.Video != nil {
-						bidType = openrtb_ext.BidTypeVideo
-					}
-					break
-				}
+			imp, err := getImpressionForBid(request.Imp, bid.ImpID)
+			if err != nil {
+				errors = append(errors, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("bid id='%s' could not find valid impid='%s'", bid.ID, bid.ImpID),
+				})
+				continue
 			}
 
-			if foundMatchingBid {
-				rv.Bids = append(rv.Bids, &adapters.TypedBid{
-					Bid:     &bid,
-					BidType: bidType,
-				})
+			bidType := openrtb_ext.BidTypeBanner
```

Comment:
I know this logic is similar to what you had before but is just extended for other media types but I wanted to ask if there's any indication of the media type in the bid that gets sent back from your server and if we can use that here rather than depending on the imp media type. The issue with the imp media type is that if the imp is a multi-format imp that is say it has both video and banner objects then this logic will always set the media type as banner when the bid returned by your server could potentially be a video

Reply:
We currently do not support multiple formats in one impression.
We don't include media types in our bid responses, but we also identify them by the value of the impression media type.
If you want I will modify my endpoints to send them to you via bidresponse.seatbid.bid[i].ext.prebid.type

I fixed the logic to not set the default Bid type to Banner.

Reply:
@thuyhq I think having your endpoint send the bid type in the bidresponse.seatbid.bid[i].ext.prebid.type is the best way to go so that in the future even if you change your endpoint to support multiple formats, this logic will still work as expected. Let us know when you've updated that on your end and can update your adapter too

Reply:
I have included the mediaType in the bid response from our endpoint and parsed the bidType from this data at the adapter.
Thank you.